sdk/helper/storagepacker/storagepacker_v2.go

```diff
@@ -0,0 +1,774 @@
+
+	shardLocks[bucket.LockEntry] = struct{}{}
+
+	shards := make(map[string]*LockedBucket, numShards)
+	for i := 0; i < numShards; i++ {
+		shardKey := fmt.Sprintf("%x", i)
+		lock := locksutil.LockForKey(s.storageLocks, cacheKey+shardKey)
+		shardedBucket := &LockedBucket{
+			LockEntry: lock,
+			Bucket: &Bucket{
+				Key:     fmt.Sprintf("%s/%s", bucket.Key, shardKey),
+				ItemMap: make(map[string][]byte),
+			},
+		}
+		shards[shardKey] = shardedBucket
+		// If it was equal we'd be locked already
+		s.Logger.Trace("created shard", "shard_key", shardKey)
+		// Don't try to lock the same lock twice in case it hashes that way
+		if _, ok := shardLocks[lock]; !ok {
+			s.Logger.Trace("locking lock", "shard_key", shardKey)
+			lock.Lock()
+			shardLocks[lock] = struct{}{}
+		}
+	}
+
+	s.Logger.Debug("resilvering items")
+
+	parentPrefix := s.GetCacheKey(bucket.Key)
+	// Resilver the items
+	for k, v := range bucket.ItemMap {
+		itemKey := strings.TrimPrefix(k, parentPrefix)[0 : s.BucketShardBits/4]
+		s.Logger.Trace("resilvering item", "parent_prefix", parentPrefix, "item_id", k, "item_key", itemKey)
+		// Sanity check
+		childBucket, ok := shards[itemKey]
+		if !ok {
+			// We didn't complete sharding so don't make other parts of the
+			// code think that it completed
+			s.Logger.Error("failed to find sharded storagepacker bucket", "bucket_key", bucket.Key, "item_key", itemKey)
+			return errors.New("failed to shard storagepacker bucket")
+		}
+		childBucket.ItemMap[k] = v
+	}
+
+	s.Logger.Debug("storing sharded buckets")
+
+	// Ensure we can write all of these buckets. Create a cleanup function if not.
+	retErr := new(multierror.Error)
+	cleanupStorage := func() {
+		for _, v := range shards {
+			if err := s.BucketStorageView.Delete(ctx, v.Key); err != nil {
+				retErr = multierror.Append(retErr, err)
+				// Don't exit out, clean up as much as possible
+			}
+		}
+	}
+	for k, v := range shards {
+		s.Logger.Trace("storing bucket", "shard", k)
+		if err := s.storeBucket(ctx, v, false); err != nil {
+			s.Logger.Debug("encountered error", "shard", k)
+			retErr = multierror.Append(retErr, err)
+			cleanupStorage()
+			return retErr
+		}
+	}
+
+	cleanupCache := func() {
+		for _, v := range shards {
+			s.bucketsCache.Delete(s.GetCacheKey(v.Key))
+		}
+	}
+	// Add to the cache. It's not too late to back out, via the cleanup cache
+	// function. We hold the lock while storing the updated original bucket so
+	// that nobody accesses it in an inconsistent state.
+	s.Logger.Debug("updating cache")
+	s.bucketsCacheLock.Lock()
+	{
+		for _, v := range shards {
+			s.bucketsCache.Insert(s.GetCacheKey(v.Key), v)
+		}
+
+		// Finally, update the original and persist
+		origBucketItemMap := bucket.ItemMap
+		bucket.ItemMap = nil
+		if err := s.storeBucket(ctx, bucket, false); err != nil {
+			retErr = multierror.Append(retErr, err)
+			bucket.ItemMap = origBucketItemMap
+			cleanupStorage()
```

Comment:
This is safe against errors in storeBucket but I'm not sure it's crash-safe. What happens if we come back up after a failure and find both the shards and the original bucket present in storage?

I think, from looking at the recovery code, that the bucketCache will be rebuilt in partial form.  If we got this far successfully, we should be fine because longest-prefix search will always put us into the correct shard, even if duplicate entries exist further up.  But, if we crashed while sharding then we could end up with:

xxx/0
xxx/1
xxx/2
xxx

So searches for xxx3 would end up in xxx but searches for xxx0 would end up in xxx/0.  This would be OK until we again attempted to reshard xxx, at which point some of the new keys in /0 /1 /2 could be overwritten.

My suggestion would be to look for incompletely-sharded buckets at recovery time and eliminate the shards.

Reply:
Good plan